aclk/agent_cloud_link.c

```diff
@@ -935,13 +936,36 @@ void *aclk_query_main_thread(void *ptr)
 // Thread cleanup
 static void aclk_main_cleanup(void *ptr)
 {
+    char payload[sizeof(ACLK_LWT_MSG)+128+1];
     struct netdata_static_thread *static_thread = (struct netdata_static_thread *)ptr;
     static_thread->enabled = NETDATA_MAIN_THREAD_EXITING;
 
     info("cleaning up...");
 
-    // Wakeup thread to cleanup
-    QUERY_THREAD_WAKEUP;
+    if (is_agent_claimed()) {
+
+        // Wakeup thread to cleanup
+        QUERY_THREAD_WAKEUP;
+        // Send a graceful disconnect message
+        time_t time_created = now_realtime_sec();
+        char *msg_id = create_uuid();
+        snprintfz(payload, sizeof(ACLK_LWT_MSG) + 128, ACLK_LWT_MSG, msg_id, time_created, "graceful");
+        aclk_send_message(ACLK_METADATA_TOPIC, payload, msg_id);
+        freez(msg_id);
+
+        _link_event_loop();
```

Comment:
Why are we calling `_link_event_loop()` three times?

Reply:
There is no indication of what is pending in the lws layer and since the event loop is not called in the main loop at that point, the only way I got this to work is by calling the event loop to make sure the message will be delivered (and the disconnect packet as well). 

Calling the event loop e.g 2 times results in the connection being dropped and the cloud gets both messages (graceful and unexpected)

The ideal would be to call the loop and peek the lws internals for pending packets.

Reply:
Would it work to check the outgoing queue size for LWS using the stats function that I put in during the stress-testing. My concern is that we tuning this for a specific code path through mosquitto and we have no idea if it will behave like that in production. I would rather see a loop (with a timeout) that keeps hitting the event loop until a condition shows that the message is gone.

Reply:
> Would it work to check the outgoing queue size for LWS using the stats function that I put in during the stress-testing. My concern is that we tuning this for a specific code path through mosquitto and we have no idea if it will behave like that in production. I would rather see a loop (with a timeout) that keeps hitting the event loop until a condition shows that the message is gone.

If that gives accurate indication that the lws layer has flushed all the packets, it will be ideal. I will integrate that into a loop.

Reply:
Added check by calling `lws_wss_check_queues` and using that information to call the event loop until the queue is empty  (timeout on 5 seconds).